### PR TITLE
enhance: alterindex & altercollection supports altering properties

### PR DESCRIPTION
**Pull request by JsDove**
enhance : 

1. alterindex delete properties
We have introduced a new parameter deleteKeys to the alterindex functionality, which allows for the deletion of properties within an index. This enhancement provides users with the flexibility to manage index properties more effectively by removing specific keys as needed.
2. altercollection delete properties
We have introduced a new parameter deleteKeys to the altercollection functionality, which allows for the deletion of properties within an collection. This enhancement provides users with the flexibility to manage collection properties more effectively by removing specific keys as needed.

3.support altercollectionfield
We currently support modifying the fieldparams of a field in a collection using altercollectionfield, which only allows changes to the max-length attribute.
Key Points:
-  New Parameter - deleteKeys: This new parameter enables the deletion of specified properties from an index. By passing a list of keys to deleteKeys, users can remove the corresponding properties from the index.

- Mutual Exclusivity: The deleteKeys parameter cannot be used in conjunction with the extraParams parameter. Users must choose one parameter to pass based on their requirement. If deleteKeys is provided, it indicates an intent to delete properties; if extraParams is provided, it signifies the addition or update of properties.

issue: https://github.com/milvus-io/milvus/issues/37436

### CODE REVIEW

**Comment by JsDove** [comment]
rerun go-sdk

**Comment by JsDove** [comment]
rerun go-sdk

File: internal/datacoord/index_service_test.go [hunk 1]
```diff
@@ -637,6 +637,41 @@ func TestServer_AlterIndex(t *testing.T) {
 		req.Params[0].Value = "true"
 	})
```
Comment (czs007) [inline internal/datacoord/index_service_test.go:639]: add a update case

File: internal/proxy/task.go [hunk 2]
```diff
@@ -980,19 +990,36 @@ func validatePartitionKeyIsolation(colName string, isPartitionKeyEnabled bool, p
...
-		if loaded {
-			return merr.WrapErrCollectionLoaded(t.CollectionName, "can not alter mmap properties if collection loaded")
+	} else if len(t.GetDeleteKeys()) > 0 {
+		if hasPropInDeletekeys(t.DeleteKeys) {
+			loaded, err := isCollectionLoaded(ctx, t.queryCoord, t.CollectionID)
+			if err != nil {
+				return err
+			}
+			if loaded {
+				return merr.WrapErrCollectionLoaded(t.CollectionName, "can not alter mmap properties if collection loaded")
```
Comment (czs007) [inline internal/proxy/task.go:1021]: can not delete properties

File: internal/rootcoord/alter_collection_task.go [hunk 3]
```diff
@@ -125,6 +133,177 @@ func (a *alterCollectionTask) Execute(ctx context.Context) error {
...
+			Base:           a.Req.Base,
+			DbName:         a.Req.DbName,
+			CollectionName: a.Req.CollectionName,
+			CollectionID:   oldColl.CollectionID,
+		},
+		core: a.core,
+	})
+
+	// properties needs to be refreshed in the cache
+	aliases := a.core.meta.ListAliasesByID(oldColl.CollectionID)
```
Comment (czs007) [inline internal/rootcoord/alter_collection_task.go:214]: unnecessary

File: internal/rootcoord/alter_collection_task.go [hunk 4]
```diff
@@ -125,6 +133,177 @@ func (a *alterCollectionTask) Execute(ctx context.Context) error {
...
+	})
+
+	oldReplicaNumber, _ := common.DatabaseLevelReplicaNumber(oldColl.Properties)
+	oldResourceGroups, _ := common.DatabaseLevelResourceGroups(oldColl.Properties)
+	newReplicaNumber, _ := common.DatabaseLevelReplicaNumber(newColl.Properties)
+	newResourceGroups, _ := common.DatabaseLevelResourceGroups(newColl.Properties)
+	left, right := lo.Difference(oldResourceGroups, newResourceGroups)
+	rgChanged := len(left) > 0 || len(right) > 0
+	replicaChanged := oldReplicaNumber != newReplicaNumber
+	if rgChanged || replicaChanged {
```
Comment (czs007) [inline internal/rootcoord/alter_collection_task.go:230]: unnecessary

**Comment by JsDove** [comment]
rerun cpp-unit-test

File: internal/proxy/task.go [hunk 5]
```diff
@@ -1065,6 +1093,89 @@ func (t *alterCollectionTask) PostExecute(ctx context.Context) error {
...
+	IsStringType := false
+	indexName := ""
+	for _, field := range collSchema.Fields {
+		if field.GetName() == t.FieldName && typeutil.IsStringType(field.DataType) {
+			IsStringType = true
+			indexName = field.GetName()
+		}
+	}
+	if !IsStringType {
+		return merr.WrapErrParameterInvalid(indexName, "it can not modify the maxlength for non-string types")
```
Comment (czs007) [inline internal/proxy/task.go:1163]: better to refine the error message.

File: internal/proxy/task.go [hunk 6]
```diff
@@ -89,6 +89,7 @@ const (
 	DescribeAliasTaskName         = "DescribeAliasTask"
 	ListAliasesTaskName           = "ListAliasesTask"
 	AlterCollectionTaskName       = "AlterCollectionTask"
+	AlterCollectionFieldTaskName  = "AlterColeectionFiledTask"
```
Comment (czs007) [inline internal/proxy/task.go:92]: typo AlterCllectionFieldTask

File: internal/rootcoord/alter_collection_task.go [hunk 7]
```diff
@@ -48,8 +49,12 @@ func (a *alterCollectionTask) Prepare(ctx context.Context) error {
 
 func (a *alterCollectionTask) Execute(ctx context.Context) error {
 	// Now we only support alter properties of collection
-	if a.Req.GetProperties() == nil {
-		return errors.New("only support alter collection properties, but collection properties is empty")
+	if a.Req.GetProperties() == nil && a.Req.GetDeleteKeys() == nil {
+		return errors.New("only support alter collection properties, but collection properties or deletekeys is empty")
```
Comment (czs007) [inline internal/rootcoord/alter_collection_task.go:53]: The collection properties to alter and keys to delete must not be empty at the same time.

File: internal/rootcoord/alter_collection_task.go [hunk 8]
```diff
@@ -125,6 +133,147 @@ func (a *alterCollectionTask) Execute(ctx context.Context) error {
...
+	props := make(map[string]string)
+	for _, prop := range oldProps {
+		props[prop.Key] = prop.Value
+	}
+	log.Info("UpdateFieldPropertyParams", zap.Any("oldprops", props), zap.Any("newprops", updatedProps))
+	for _, prop := range updatedProps {
+		props[prop.Key] = prop.Value
+	}
+	log.Info("UpdateFieldPropertyParams", zap.Any("newprops", props))
+	propKV := make([]*commonpb.KeyValuePair, 0)
```
Comment (czs007) [inline internal/rootcoord/alter_collection_task.go:266]: propKV := make([]*commonpb.KeyValuePair, 0, len(props))

File: internal/rootcoord/alter_collection_task.go [hunk 9]
```diff
@@ -125,6 +133,147 @@ func (a *alterCollectionTask) Execute(ctx context.Context) error {
...
+}
+
+func UpdateFieldProperties(coll *model.Collection, fieldName string, updatedProps []*commonpb.KeyValuePair) error {
+	for i, field := range coll.Fields {
+		if field.Name == fieldName {
+			coll.Fields[i].TypeParams = UpdateFieldPropertyParams(field.TypeParams, updatedProps)
+			return nil
+		}
+	}
+	return merr.WrapErrParameterInvalidMsg("%s does not exist in collection", fieldName)
```
Comment (czs007) [inline internal/rootcoord/alter_collection_task.go:253]: field %s does not exist ……

**Comment by JsDove** [comment]
rerun go-sdk

**Comment by JsDove** [comment]
rerun go-sdk

**Comment by JsDove** [comment]
rerun ut

**Comment by JsDove** [comment]
rerun go-sdk

**Pull request merged**